File: old/src/scanners/PPSSPPSource.js
const fsp = require("fs/promises");
const path = require("path");

const GameDir = require("./GameDir.js");
const config = require("../utils/configFormats.js");

const EmulationSource = require("./EmulationSource.js");
const PPSSPPGame = require("../games/PPSSPPGame");

const USER_DIR = process.env["HOME"];

class PPSSPPSource extends EmulationSource {

	static name = "PPSSPP";
	static gameClass = PPSSPPGame;

	preferCache = false;

	configPath = `${USER_DIR}/.config/ppsspp/PSP/SYSTEM/ppsspp.ini`;
	romRegex = /.+\.(iso|cso)/i;

	constructor() {
		super();
	}

	/**
	 * Get ppsspp config data from it config file in $HOME/.config/ppsspp/PSP/SYSTEM/ppsspp.ini
	 * @returns {object} - The config data
	 */
	async _getConfig() {

		const configFileContents = await fsp.readFile(this.configPath, "utf-8");
		const configData = config.config2js(configFileContents);
		return configData;

	}

	/**
	 * Get ppsspp ROM dirs from its config data (pinned paths)
	 * @param {object} configData - ppsspp config data
	 * @returns {GameDir[]} - The game dirs extracted from ppsspp's config data
	 * @private
	 */
	async _getROMDirs(configData) {

		const dirs = [];
		const pathsObj = configData?.["PinnedPaths"];
		if (!pathsObj) {
			return dirs;
		}
		const paths = Object.values(pathsObj);
		for (const dirPath of paths) {
			dirs.push(new GameDir(dirPath, false));
		}

		return dirs;

	}

	/**
	 * Get all ppsspp ROMs in the specified game dirs
	 * @param {GameDir[]} dirs - The game dirs to scan for ROMs
	 * @returns {PPSSPPGame[]} - An array of found games
	 * @private
	 */
	async _getROMGames(dirs) {

		const gamePaths = await this._getROMs(dirs, this.romRegex);
		const games = [];
		for (const gamePath of gamePaths) {
			const game = new this.constructor.gameClass(path.basename(gamePath), gamePath);
			game.isInstalled = true;
			games.push(game);
		}

		return games;

	}

	/**
	 * Get all ppsspp games
	 * @returns {PPSSPPGame[]} - An array of found games
	 */
	async scan() {
		const config = await this._getConfig();
		const romDirs = await this._getROMDirs(config);
		const romGames = await this._getROMGames(romDirs);
		return romGames;
	}

}

module.exports = PPSSPPSource;

File: gali/src/sources/ppsspp/ppsspp_source.py
from pathlib import PurePath

from gali.utils.explicit_config_parser import ExplicitConfigParser
from gali.sources.emulation_source import EmulationSource
from gali.sources.game_dir import GameDir
from gali.sources.ppsspp.ppsspp_game import PPSSPPGame
from gali.sources.file_dependent_source import FileDependentSource


class PPSSPPSource(EmulationSource, FileDependentSource):

    name: str
    config_path: str
    game_class: type[PPSSPPGame]
    
    rom_extensions: tuple[str] = (".iso", ".cso")

    def get_config(self) -> ExplicitConfigParser:
        config = ExplicitConfigParser()
        config.read(self.config_path, encoding="utf-8-sig")
        return config

    def get_rom_dirs(self, config: ExplicitConfigParser) -> tuple[GameDir]:
        rom_dirs = []
        items = config.items("PinnedPaths")
        for (key, path) in items:
            rom_dirs.append(GameDir(path, 0))
        return tuple(rom_dirs)

    def get_rom_games(self, rom_dirs: tuple[GameDir]) -> tuple[PPSSPPGame]:
        games = []
        for rom_dir in rom_dirs:
            try:
                rom_paths = self.get_rom_paths(rom_dir, self.rom_extensions)
            except OSError:
                continue
            for path in rom_paths:
                name = PurePath(path).name
                game = self.game_class(
                    name=name,
                    game_path=path,
                    is_installed=True,
                )
                games.append(game)
        return tuple(games)

    def scan(self) -> list[PPSSPPGame]:
        config = self.get_config()
        rom_dirs = self.get_rom_dirs(config)
        rom_games = self.get_rom_games(rom_dirs)
        return rom_games

    def get_precondition_file_path(self):
        return self.config_path